_specifications/specification-3-17.md

```diff
@@ -2419,6 +2426,14 @@ interface ServerCapabilities {
 		}
 	}
 
+	/**
+	 * The server provides type hierarchy support.
+	 *
+	 * @since 3.17.0
+	 */
+	typeHierarchyProvider?: boolean | TypeHierarchyOptions
```

Comment:
Is a simple `boolean` necessary/sufficient here since each server capability can be defined under `TypeHierarchyOptions`?

Reply:
Yeah, we don't need `boolean` here in this version. We are considering whether to remove the capability `inheritanceTreeSuppport` to keep the type hierarchy general. I'll update this PR then.

Reply:
I find it was referring call hierarchy.
```ts
callHierarchyProvider?: boolean | CallHierarchyOptions | CallHierarchyRegistrationOptions;
```

which is equivalent to: 
``` ts
boolean // simplest 
OR
{
  workDoneProgress?: boolean;  // from WorkDoneProgressOptions
} 
OR
{
  workDoneProgress?: boolean;  // from WorkDoneProgressOptions
  id?: string;  // from StaticRegistrationOptions 
  documentSelector: DocumentSelector | null;  // from TextDocumentRegistrationOptions 
}
```

specifying whether server has capabitility to support work done progress / unregistration / customized scope of the feature.


> Is a simple boolean necessary/sufficient here since each server capability can be defined under TypeHierarchyOptions?

@KamasamaK  I think `boolean` is **insufficient** for advanced features mentioned above. For me, it's somehow **necessary** . Because on one hand, when I implement it in server side I might not care about any advanced feature, then I can simply return a `true`; on the other hand it's consistent with other capabilities like call hierarchy.

> We are considering whether to remove the capability inheritanceTreeSuppport to keep the type hierarchy general.

@CsCherrYY  I'm voting +1 for removing `inheritanceTreeSuppport`, otherwise for languages supporting multiple inheritance (e.g. C++), you may also need something like `inheritanceGraphSuppport`?

Reply:
If you don't care about implementing or can't implement advanced features, you can return `false` for their capabilities. For example, `codeLensProvider` only has a type of `CodeLensOptions` with one optional native capability, so that's already established for LSP. That is the reason I say it is probably unnecessary. But if the extra capability is going to be removed, that might change things.

Perhaps @dbaeumer can weigh in on the necessity of a simple `boolean` for new operations when *Options exists, and whether them having no or only optional native properties makes a difference in that determination.

Reply:
Thank you for pointing out `CodeLensOptions`. It looks:
`{ resolveProvider: false }`  means server doesn't have a resolve provider.
`{ resolveProvider: true }`  means server also support to resolve code lens.
So now I have a question, if server doesn't support code lens at all, what should it return? (if boolean is not allowed)

Reply:
Excluding `codeLensProvider` would indicate that the server does not support it.

> A missing property should be interpreted as an absence of the capability. If a missing property normally defines sub properties, all missing sub properties should be interpreted as an absence of the corresponding capability.